### PR TITLE
Fix MvStudentT logp

### PR DESCRIPTION
**Pull request by aseyboldt**
The density for `MvStudentT` used to be wrong for multiple observations. This should fix that and also add support for using the cholesky decomposition in MvStudentT in the same way as in `MvNormal`.

### CODE REVIEW

**Comment by twiecki** [comment]
That's a nice refactor + fix. I suppose the Cholesky decomposition is just copied over from the MvNormal? Was the MvStudentT bug anything concrete or just a random math / implementation bug?

**Comment by aseyboldt** [comment]
I moved most of the common stuff between mvnormal mvstudent to a common base class, so there wasn't much copying going on. 🙂
The bug was that `(x - mu).T S (x - mu)` wasn't right for the vector-valued x. It was
```
(x - mu).T @ solve(S, x - mu)
```
but it should have been
```
((x - mu) * solve(S, x - mu)).sum(-1)
```

**Comment by twiecki** [comment]
```
_______________________________ test_repr_latex_ _______________________________
    def test_repr_latex_():
        with Model():
            x0 = Binomial('Discrete', p=.5, n=10)
            x1 = Normal('Continuous', mu=0., sd=1.)
            x2 = GaussianRandomWalk('Timeseries', mu=x1, sd=1., shape=2)
            x3 = MvStudentT('Multivariate', nu=5, mu=x2, Sigma=np.diag(np.ones(2)), shape=2)
            x4 = NormalMixture('Mixture', w=np.array([.5, .5]), mu=x3, sd=x0)
    
        assert x0._repr_latex_() == '$Discrete \\sim \\text{Binomial}' \
                                    '(\\mathit{n}=10, \\mathit{p}=0.5)$'
        assert x1._repr_latex_() == '$Continuous \\sim \\text{Normal}' \
                                    '(\\mathit{mu}=0.0, \\mathit{sd}=1.0)$'
        assert x2._repr_latex_() == '$Timeseries \\sim \\text' \
                                    '{GaussianRandomWalk}(\\mathit{mu}=Continuous, ' \
                                    '\\mathit{sd}=1.0)$'
>       assert x3._repr_latex_() == '$Multivariate \\sim \\text{MvStudentT}' \
                                    '(\\mathit{nu}=5, \\mathit{mu}=Timeseries, ' \
                                    '\\mathit{Sigma}=array)$'
pymc3/tests/test_distributions.py:947: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymc3/model.py:894: in _repr_latex_
    return self.distribution._repr_latex_(name=name, dist=dist)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pymc3.distributions.multivariate.MvStudentT object at 0x7f2bb9fb8050>
name = 'Multivariate'
dist = <pymc3.distributions.multivariate.MvStudentT object at 0x7f2bb9fb8050>
    def _repr_latex_(self, name=None, dist=None):
        if dist is None:
            dist = self
        mu = dist.mu
        nu = dist.nu
>       Sigma = dist.Sigma
E       AttributeError: 'MvStudentT' object has no attribute 'Sigma'
pymc3/distributions/multivariate.py:356: AttributeError
```

**Comment by twiecki** [comment]
```
_____________________ test_fit_oo[FullRankADVI-full-scale] _____________________
inference = <pymc3.variational.inference.FullRankADVI object at 0x7f0e1b557f90>
fit_kwargs = {'n': 6000, 'obj_optimizer': <functools.partial object at 0x7f0e1ae80050>}
simple_model_data = {'d': 111.36111111111111, 'data': array([ -3.50985754e+00,  -5.41479290e+00,  -3.05693439e+00,
        -4.309104...  3...581e-01,  -3.07747142e+00,  -6.71353697e+00,
        -3.28225166e+00]), 'mu0': 4.0, 'mu_post': -4.921929491177873, ...}
    @pytest.mark.run('first')
    def test_fit_oo(inference,
                    fit_kwargs,
                    simple_model_data):
>       trace = inference.fit(**fit_kwargs).sample(10000)
pymc3/tests/test_variational_inference.py:183: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymc3/variational/inference.py:126: in fit
    step_func = self.objective.step_function(score=score, **kwargs)
pymc3/memoize.py:16: in memoizer
    cache[key] = obj(*args, **kwargs)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/configparser.py:117: in res
    return f(*args, **kwargs)
pymc3/variational/opvi.py:247: in step_function
    total_grad_norm_constraint=total_grad_norm_constraint)
pymc3/variational/opvi.py:148: in updates
    total_grad_norm_constraint=total_grad_norm_constraint
pymc3/variational/opvi.py:178: in add_obj_updates
    obj_target = self(obj_n_mc, more_obj_params=more_obj_params)
pymc3/variational/opvi.py:296: in __call__
    a = self.op.apply(self.tf)
pymc3/variational/operators.py:24: in apply
    return self.logq_norm - self.logp_norm
pymc3/variational/opvi.py:333: in <lambda>
    logq_norm = property(lambda self: self.approx.logq_norm)
pymc3/memoize.py:16: in memoizer
    cache[key] = obj(*args, **kwargs)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/configparser.py:117: in res
    return f(*args, **kwargs)
pymc3/variational/opvi.py:1020: in logq_norm
    return self.logq / self.normalizing_constant
pymc3/memoize.py:16: in memoizer
    cache[key] = obj(*args, **kwargs)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/configparser.py:117: in res
    return f(*args, **kwargs)
pymc3/variational/opvi.py:1016: in logq
    return self.symbolic_log_q_W.mean(0)
pymc3/memoize.py:16: in memoizer
    cache[key] = obj(*args, **kwargs)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/configparser.py:117: in res
    return f(*args, **kwargs)
pymc3/variational/opvi.py:1009: in symbolic_log_q_W
    q_w_global = self.symbolic_log_q_W_global
pymc3/memoize.py:16: in memoizer
    cache[key] = obj(*args, **kwargs)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/configparser.py:117: in res
    return f(*args, **kwargs)
pymc3/variational/approximations.py:217: in symbolic_log_q_W_global
    return pm.MvNormal.dist(mu=mu, chol=L).logp(z)
pymc3/distributions/multivariate.py:273: in logp
    norm = - 0.5 * k * pm.floatX(np.log(2 * np.pi))
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/tensor/var.py:233: in __rmul__
    return theano.tensor.basic.mul(other, self)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/op.py:615: in __call__
    node = self.make_node(*inputs, **kwargs)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/tensor/elemwise.py:581: in make_node
    out_broadcastables)]
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/type.py:405: in __call__
    return utils.add_tag_trace(self.make_variable(name))
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/tensor/type.py:353: in make_variable
    return self.Variable(self, name=name)
```

**Comment by twiecki** [comment]
Seems like there are still some float32 issues.

**Comment by aseyboldt** [comment]
That one was hiding well, but I think I finally found it.

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -349,11 +349,11 @@ def random(self, point=None, size=None):
 
     def logp(self, value):
         quaddist, logdet, ok = self._quaddist(value)
-        k = value.shape[-1]
+        k = value.shape[-1].astype(theano.config.floatX)
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: subtle!

**Comment by twiecki** [comment]
LGTM, thanks for the fix.

**Pull request merged**